evoting/lib/election.go

```diff
@@ -53,94 +58,56 @@ type footer struct {
 	ContactEmail string // ContactEmail stores the email address of the Contact person.
 }
 
-func init() {
-	network.RegisterMessages(Election{}, Ballot{}, Box{}, Mix{}, Partial{})
-}
-
-// FetchElection retrieves the election object from its skipchain and sets its stage.
-func FetchElection(roster *onet.Roster, id skipchain.SkipBlockID) (*Election, error) {
-	chain, err := chain(roster, id)
+// GetElection fetches the election structure from its skipchain and sets the stage.
+func GetElection(roster *onet.Roster, id skipchain.SkipBlockID) (*Election, error) {
+	client := skipchain.NewClient()
+	reply, err := client.GetUpdateChain(roster, id)
 	if err != nil {
 		return nil, err
 	}
 
-	_, blob, _ := network.Unmarshal(chain[1].Data, cothority.Suite)
-	election := blob.(*Election)
+	transaction := UnmarshalTransaction(reply.Update[1].Data)
+	if transaction == nil || transaction.Election == nil {
+		return nil, fmt.Errorf("no election structure in %s", id.Short())
+	}
+	election := transaction.Election
 
-	n, numMixes, numPartials := len(election.Roster.List), 0, 0
-	for _, block := range chain {
-		_, blob, _ := network.Unmarshal(block.Data, cothority.Suite)
-		if _, ok := blob.(*Mix); ok {
-			numMixes++
-		} else if _, ok := blob.(*Partial); ok {
-			numPartials++
+	n, mixes, partials := len(election.Roster.List), 0, 0
+	for _, block := range reply.Update {
```

Comment:
This will break with a non-1 `BaseHeight` as proposed above, because `GetUpdateChain` will return the _shortest_ path from the genesis to the last block.

Furthermore, this might take quite some time if we have 1000s of votes on the skipchain. My proposition is to store that information in the service and save it every time it is updated (what we do in most other services). Even though it means that the service needs to hold state then.

For a future version, we'll be able to use the skipchains in a more database-like way and have the whole state available in a provable way at every block. But that's for later.

Also think about what to do with the other `GetUpdateChain` calls you have in the code.

Reply:
Keeping track of the election stage turned out to be more complicated than I initially thought. Especially now with the transaction verification function that is executed on all nodes. There are
cases when it is not clear. E.g. during protocols the leaf nodes never know if the protocol has finished correctly.
I decided to drop the field altogether. The front-end can also use the getter messages like GetMixes and GetPartials to determine the election stage.
The shared secret however is now stored on disk. This fixes #1092.